docs/development/best-practices.md

```diff
@@ -0,0 +1,125 @@
+# Best practices
+
+## General
+
+- Prefer full function declarations for readability and better stack traces, so avoid `const func = ():void => {}`
+- Avoid [Enums](https://github.com/renovatebot/renovate/issues/13743), use union or [immutable objects](https://github.com/renovatebot/renovate/blob/5043379847818ac1fa71ff69c098451975e95710/lib/modules/versioning/pep440/range.ts#L8-L20) instead
+- Always add unit tests for full code coverage
+  - Only use `istanbul` comments for unreachable code coverage that is needed for `codecov` completion
+  - Use descriptive `istanbul` comments
+
+```js
+// istanbul ignore next: can never happen
+```
+
+### Logging
+
+- Use logger metadata, like this:
+
+```js
+logger.trace({ obj }, 'title');
+```
+
+Avoid:
+
+```js
+logger.trace('title:\n' + obj);
+```
+
```

Comment:
```suggestion
Use logger metadata if logging for `WARN`, `ERROR`, `FATAL`, or if the result is a complex metadata object needing a multiple-line pretty stringification.
Otherwise, inline metadata into the log message if logging at `INFO` or below, or if the metadata object is complex.

`WARN` and above messages are often used in metrics or error catching services, and should have a consistent `msg` component so that they will be automatically grouped/associated together.
Metadata which is separate from its message is harder for human readability, so try to combine it in the message unless it's too complex to do so.

Good:

```js
logger.debug({ config }, 'Full config');
logger.warn({ presetName }, 'Failed to look up preset');
```

Avoid:

```js
logger.debug({ branchName }, 'Generated branchName');
logger.warn(`Failed to look up preset ${presetName}`);
```

```

Reply:
@JamieMagee @viceice I've never really brought this one up before btw, but it's come to mind lots when I'm trying to manage the app at scale, use Sentry for grouping warnings, etc.

Reply:
Makes sense, but the different advice for INFO would make this hard to enforce with a lint rule.

Reply:
I'm happy to only enforce for warn and up if possible

Reply:
I like the new text, but there's something wrong with the suggestion.

I find working with review suggestions on GitHub hard when there are multiple codeblocks. It always seems to get messed up somehow... 😄

Reply:
the suggestion needs to have more backticks than the code blocks to get it working. 😉

Reply:
I'll commit it then fix it